### PR TITLE
Add an Istiod-remote chart

### PR DESCRIPTION
**Pull request by linsun**
Add an Istiod-remote part, as part of https://github.com/istio/istio/issues/22934, we found out that the remote plane would need to have at the minimum:
- mutating WebHooks configuration that can point to central istiod
- sidecar injector config so that sidecar injection can work properly with `istioctl kube-inject` cmd towards the remote plane.

### CODE REVIEW

**Comment by linsun** [comment]
cc @sdake @howardjohn @costinm

File: manifests/charts/istiod-remote/files/injection-template.yaml [hunk 1]
```diff
@@ -0,0 +1,414 @@
+template: |
```
Comment (howardjohn) [inline manifests/charts/istiod-remote/files/injection-template.yaml:1]: I don't think this makes sense here. Its confusing - users will think injection is happening on this cluster, when it isn't actually. `istioctl` does not actually require it - it can be configured to fetch from the remote cluster or from files. I suspect if we actually did do this and people were doing kube-inject they would often find this to be out of sync with the remote cluster and see odd behavior. Additionally, its hard to maintain.

File: manifests/charts/istiod-remote/templates/istiod-injector-configmap.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+{{- if not .Values.global.omitSidecarInjectorConfigMap }}
```
Comment (howardjohn) [inline manifests/charts/istiod-remote/templates/istiod-injector-configmap.yaml:1]: same comment about the injection template

File: manifests/charts/istiod-remote/templates/mutatingwebhook.yaml [hunk 3]
```diff
@@ -0,0 +1,66 @@
+# Installed for each revision - not installed for cluster resources ( cluster roles, bindings, crds)
```
Comment (howardjohn) [inline manifests/charts/istiod-remote/templates/mutatingwebhook.yaml:1]: If we make the suggestions above this is just a single object. What benefit do we get making a complete helm chart rather than just a yaml file under samples/?

File: manifests/charts/istiod-remote/files/injection-template.yaml (hunk 1 above)
Comment (linsun) [reply manifests/charts/istiod-remote/files/injection-template.yaml:1]: istioctl Kube-inject currently reads the default injection info from the cluster where user is running by default.  Are you saying we should configure the sidecar injector cm to read the injection template from the main cluster?  I think requiring not having istioctl Kube-inject working out of box and require configuring params is unintuitive.   From my testing, the istio sidecar injector cm needs to be in place in the cluster where user runs istioctl Kube-inject to get the injection info.

Comment (howardjohn) [reply manifests/charts/istiod-remote/files/injection-template.yaml:1]: It does work out of the box, at least in my opinion. 

With current istioctl:
* If I want to inject my pod using cluster1, I run `istioctl kube-inject --kubeconfig cluster1`.
* If I want to inject my pod using cluster2, I run `istioctl kube-inject --kubeconfig cluster2`.
* If I want to use files (recommended) I pass `--injectConfigFile`

In the remote cluster model, the same model remains. Say our kubeconfigs are `local` and `remote`. User would run `istioctl kube-inject --kubeconfig remote | kubectl apply --kubeconfig local`. This is accurately modeling what is happening - the remote cluster is injecting the pod, and its deployed in the local cluster, just like using auto injections.

File: manifests/charts/istiod-remote/templates/mutatingwebhook.yaml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+{{- end }}
+  labels:
+    istio.io/rev: {{ .Values.revision | default "default" }}
+    app: sidecar-injector
+    release: {{ .Release.Name }}
+webhooks:
+  - name: sidecar-injector.istio.io
+    clientConfig:
+      url: https://{{ .Values.sidecarInjectorWebhook.remoteWebhookHost }}:{{ .Values.sidecarInjectorWebhook.remoteWebhookPort }}/inject
+      caBundle: ""
```
Comment (morvencao) [inline manifests/charts/istiod-remote/templates/mutatingwebhook.yaml:19]: Need to figure out how to retrieve the CA bundle to validate the webhook's server certificate.

Istiod will patch the `MutatingWebhookConfiguration` with CA bundle before starting if `MutatingWebhookConfiguration` is in the same cluster.

How to patch it if the `MutatingWebhookConfiguration` in different cluster, will be some manual steps needed?

Comment (linsun) [reply manifests/charts/istiod-remote/templates/mutatingwebhook.yaml:19]: Plan to modify the WebHooks code to patch it when centralIstiod is enabled.

File: manifests/charts/istiod-remote/files/injection-template.yaml (hunk 1 above)
Comment (linsun) [reply manifests/charts/istiod-remote/files/injection-template.yaml:1]: What if user doesn't have kubeconfig for the central istiod?  e.g. I'm dedicated a central istiod cluster for just running istio control plane.   Normal user can only access the remote control plane cluster.  In this case, I can't ask users to do

```istioctl kube-inject --kubeconfig remote | kubectl apply --kubeconfig local```

Comment (howardjohn) [reply manifests/charts/istiod-remote/files/injection-template.yaml:1]: Then what makes you think the `injection-template.yaml` in your cluster actually matches what is used in the remote cluster if you don't control the remote cluster? And why would you want to use the configmap rather than files?

Comment (linsun) [reply manifests/charts/istiod-remote/files/injection-template.yaml:1]: Chatted with @howardjohn offline to understand his concerns.  Basically the chart is just to lay off the initial configuration, we expect central istiod to patch/refresh the WebHooks and configmaps (Istio-sidecar-injector and possibly istio-ca-root-cert) as necessary.

**Review by howardjohn — COMMENTED** [review]
If we are going to have the injection template in here, my main concern is keeping it in sync. Can we have a make target (as a dependency of `gen`) that copy+pastes the main on here? I think we cannot use literal symlinks so basically a manual symlink

**Comment by linsun** [comment]
@costinm @howardjohn think all comments are addressed - can you look and approve?

File: Makefile.core.mk [hunk 5]
```diff
@@ -356,18 +356,25 @@ refresh-goldens:
 
 update-golden: refresh-goldens
 
-gen: go-gen mirror-licenses format update-crds operator-proto gen-kustomize update-golden
+gen: go-gen mirror-licenses format update-crds operator-proto copy-from-istiod gen-kustomize update-golden
```
Comment (howardjohn) [inline Makefile.core.mk:359]: nit: trailing space

File: Makefile.core.mk [hunk 6]
```diff
@@ -356,18 +356,25 @@ refresh-goldens:
...
+copy-from-istiod:
+	cp manifests/charts/istio-control/istio-discovery/files/injection-template.yaml manifests/charts/istiod-remote/files/
+	cp manifests/charts/istio-control/istio-discovery/templates/istiod-injector-configmap.yaml manifests/charts/istiod-remote/templates/
+
 # Generate kustomize templates.
 gen-kustomize:
 	helm3 template istio --include-crds manifests/charts/base > manifests/charts/base/files/gen-istio-cluster.yaml
 	helm3 template istio --namespace istio-system manifests/charts/istio-control/istio-discovery \
 		-f manifests/charts/global.yaml > manifests/charts/istio-control/istio-discovery/files/gen-istio.yaml
+	helm3 template istiod-remote manifests/charts/istiod-remote \
```
Comment (howardjohn) [inline Makefile.core.mk:376]: I don't think we need this unless you are planning to use kustomize?

File: Makefile.core.mk [hunk 7]
```diff
@@ -356,18 +356,25 @@ refresh-goldens:
...
-gen: go-gen mirror-licenses format update-crds operator-proto gen-kustomize update-golden
+gen: go-gen mirror-licenses format update-crds operator-proto copy-from-istiod gen-kustomize update-golden 
 
 check-no-modify:
 	@bin/check_no_modify.sh
 
 gen-check: check-no-modify gen check-clean-repo
 
+# Copy the injection template file from istiod chart to istiod-remote chart
+copy-from-istiod:
```
Comment (nmittler) [inline Makefile.core.mk:367]: Sorry for the drive-by review ... can we rename this target? The current name is non-descriptive. Maybe something like `istiod-remove-charts`?

File: .gitignore [hunk 8]
```diff
@@ -65,6 +65,8 @@ install/kubernetes/istio-init.yaml
 install/kubernetes/istio-non-mcp.yaml
 install/kubernetes/istio-minimal.yaml
 install/kubernetes/helm/istio/requirements.lock
+manifests/charts/istiod-remote/files/injection-template.yaml
```
Comment (howardjohn) [inline .gitignore:68]: How are these used if they are in gitignore? They won't be published anywhere then, is that intended?

**Review by howardjohn — COMMENTED** [review]
Overall looks good, would like to have someone else have eyes on this as well though since its a whole new chart

File: Makefile.core.mk (hunk 7 above)
Comment (linsun) [reply Makefile.core.mk:367]: np!  this doesn't generate istiod remote chart... it is just copy 2 files over.  generate the file is part of the gen-kustomize

File: Makefile.core.mk (hunk 6 above)
Comment (linsun) [reply Makefile.core.mk:376]: possibly we may

File: .gitignore (hunk 8 above)
Comment (linsun) [reply .gitignore:68]: basically these files are generated as part of make gen.  Having them here will ensure we don't accidentally check them in git.  remember you requested them not in git!

File: Makefile.core.mk (hunk 5 above)
Comment (linsun) [reply Makefile.core.mk]: updated

File: .gitignore (hunk 8 above)
Comment (howardjohn) [reply .gitignore:68]: How are you going to use it though? I assume some user would download the istio tar.gz and run `helm install charts/istiod-remote` but these files will be missing?

Comment (linsun) [reply .gitignore:68]: oh :) that is a good point.  Wouldn't the istio tar.gz include the `make gen` artifact?   Perhaps that is my wishful thinking.

Comment (linsun) [reply .gitignore:68]: ok I will check them in.  It will work when user clones the repo as well without running `make gen`

File: Makefile.core.mk (hunk 7 above)
Comment (sdake) [reply Makefile.core.mk:367]: `gen-sync-remote-templates` could be a better make target.

File: Makefile.core.mk (hunk 6 above)
Comment (sdake) [reply Makefile.core.mk:376]: It appears from this review there are two use cases `kustomize` and separately `helm3`.

File: manifests/charts/istiod-remote/Chart.yaml [hunk 9]
```diff
@@ -0,0 +1,13 @@
+apiVersion: v1
```
Comment (sdake) [inline manifests/charts/istiod-remote/Chart.yaml:1]: This directory is better named `istio-remote` as to be consistent with the rest of the charts in `$ISTIO/istio/manifest`. `istiod` is an internal implementation detail, and its better that we don't expose it to users, in the unlikely event we change our architecture in the future.

File: manifests/charts/istiod-remote/Chart.yaml [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+name: istiod-remote
+version: 1.1.0
```
Comment (sdake) [inline manifests/charts/istiod-remote/Chart.yaml:3]: please change the version to 1.6.0. We have, or had at one point, as part of the release process an automated sed of version and appVersion fields from 1.6.0 to the latest release.

File: manifests/charts/istiod-remote/Chart.yaml [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+appVersion: 1.1.0
+tillerVersion: ">=2.7.2"
+description: Helm chart for istio control plane
+keywords:
+  - istio
+  - istiod-remote
```
Comment (sdake) [inline manifests/charts/istiod-remote/Chart.yaml:9]: can you call this keyword `istio-remote`.

File: manifests/charts/istiod-remote/NOTES.txt [hunk 12]
```diff
@@ -0,0 +1,3 @@
+Minimal istiod for Istio remote data plane. Mutating webhook configuration and sidecar injector configuration are included.
```
Comment (sdake) [inline manifests/charts/istiod-remote/NOTES.txt:1]: `Minimal Istio deployment for remote data plane including a mutating webhook configuration and sidecar injection configuration.`

File: manifests/charts/istiod-remote/files/gen-istiod-remote.yaml [hunk 13]
```diff
@@ -0,0 +1,686 @@
+---
```
Comment (sdake) [inline manifests/charts/istiod-remote/files/gen-istiod-remote.yaml:1]: autogenerated code.

File: manifests/charts/istiod-remote/files/gen-istiod-remote.yaml [hunk 14]
```diff
@@ -0,0 +1,686 @@
...
+  labels:
+    istio.io/rev: default
+    release: istiod-remote
+data:
+
+  values: |-
+    {
+      "global": {
+        "arch": {
+          "amd64": 2,
```
Comment (sdake) [inline manifests/charts/istiod-remote/files/gen-istiod-remote.yaml:17]: a note for other reviewers, at some point arm enablement must be done for the arch selection - if this field is currently used.

File: manifests/charts/istiod-remote/files/injection-template.yaml (hunk 1 above)
Comment (sdake) [reply manifests/charts/istiod-remote/files/injection-template.yaml:1]: @howardjohn I am unclear if you are satisfied with this answer. If so, please click the resolve conversation button.

File: manifests/charts/istiod-remote/templates/istiod-injector-configmap.yaml (hunk 2 above)
Comment (sdake) [reply manifests/charts/istiod-remote/templates/istiod-injector-configmap.yaml:1]: same comment about resolve. I understand how much simpler this would be to deliver as a file in the `samples` directory.

**Review by sdake — COMMENTED** [review]
Hi Gang. This PR looks suitable - with a few minor nits, for merge. The points raised in the review assume an operator won't wish to use `helm` tooling to install the remote. There is also an assumption that the `operator` codebase won't be used to install the remote. While I understand these things don't exist today, I believe these two things would make sense for some set of use cases, and delivering this remote chart to our end users for compatibility with tooling they already have makes sense.

If anyone has objections to this approach, please raise them prior to the final freeze.

Cheers
-steve

File: manifests/charts/istiod-remote/Chart.yaml (hunk 9 above)
Comment (linsun) [reply manifests/charts/istiod-remote/Chart.yaml:1]: I have issue to name it Istio-remote because we will ask users to install the base charts in addition to Istiod-remote.  So it would be confusion if we call it Istio-remote.

File: manifests/charts/istiod-remote/Chart.yaml (hunk 10 above)
Comment (linsun) [reply manifests/charts/istiod-remote/Chart.yaml:3]: Can we leave this version thing for another issue discussion?  the main istiod chart is using 1.1.0 so I don't want to have have remote move to 1.6.0. :)

File: Makefile.core.mk (hunk 7 above)
Comment (linsun) [reply Makefile.core.mk]: ok 2 people have issue with this name, I've renamed :)

File: manifests/charts/istiod-remote/Chart.yaml (hunk 11 above)
Comment (linsun) [reply manifests/charts/istiod-remote/Chart.yaml:9]: see comment earlier

File: manifests/charts/istiod-remote/NOTES.txt (hunk 12 above)
Comment (linsun) [reply manifests/charts/istiod-remote/NOTES.txt]: updated

File: manifests/charts/istiod-remote/Chart.yaml (hunk 10 above)
Comment (sdake) [reply manifests/charts/istiod-remote/Chart.yaml:3]: @howardjohn I don't recall how you handled the chart versioning in the release of the manifests. In the past, we had a sed operation that did this work. I even wrote the sed - a year or more ago. That was replaced by the genbuilder tool. I am not clear how or if genbuilder modifies the output to version the charts.

**Comment by sdake** [comment]
@howardjohn the chart looks fine to me to merge. I am unclear on one point - the versioning of the chart, and hope you can offer an easy answer on whether the version number modified as part of the release builder process.

**Review by sdake — APPROVED** [review]
